### PR TITLE
[Testing] Support Basic Auth for multiple gateways

### PR DESCRIPTION
**Pull request by stefanprodan**
## Description

This PR adds a new command named `login` that allows users to register multiple gateways protected with basic auth. The credentials are being saved to `$HOME/.openfaas/config.yml` on Linux/macOS and `%USERPROFILE%/.openfaas/config.yml` on Windows. The `yml` extension is used to avoid overriding the openfaas-bitbar `config.yaml`. 

CLI `login` command specs:

* validates the credentials by calling `list` for a given gateway
* issues a warning if the password is passed via command line and advise on using stdin
* issues a warning if the gateway is not on HTTPS and advise on using Letsencrypt
* creates or updates credentials for a given gateway
* encodes the username and password with base64 for local storage
* sets auth to `basic` type in the YAML auth store

Upon calling a gateway API endpoint, the config store lookups a matching gateway URL entry. If an entry is found, the username and password are being decoded from base64 and used to setup the basic auth header. All API calls that return a 401 error will point the user to the `login` command.

CLI `login` command help:

```
./faas-cli login -h
Log in to OpenFaaS gateway.
If no gateway is specified, the default local one will be used.

Usage:
  faas-cli login [--username USERNAME] [--password PASSWORD] [--gateway GATEWAY_URL] [flags]

Examples:
  faas-cli login -u user -p password --gateway http://localhost:8080
  cat ~/faas_pass.txt | faas-cli login -u user --password-stdin --gateway https://openfaas.mydomain.com

Flags:
      --gateway string    Gateway URI (default "http://localhost:8080")
  -h, --help              help for login
  -p, --password string   Gateway password
      --password-stdin    Reads the gateway password from stdin
  -u, --username string   Gateway username
```

CLI `logout` command help:

```
./faas-cli logout --help
Log out from OpenFaaS gateway.
If no gateway is specified, the default local one will be used.

Usage:
  faas-cli logout [--gateway GATEWAY_URL] [flags]

Examples:
  faas-cli logout --gateway https://openfaas.mydomain.com

Flags:
      --gateway string   Gateway URI (default "http://localhost:8080")
  -h, --help             help for logout
```

## Motivation and Context
- [x] I have raised an issue to propose this change (see #178)

## How Has This Been Tested?

Tested against OpenFaaS (gateway v0.6.5) and Caddy as reverse proxy with basic auth and TLS. 

Gateway Docker Swarm test environments:
* GCP Compute Engine
* Ubuntu 16.04
* Docker Swarm CE 17.09.0-ce and Docker Swarm EE 17.06.2-ee-3
* Caddy 0.10.9 [FOSS edition](https://hub.docker.com/r/stefanprodan/caddy/)
* Letsencrypt certificate

Gateway Kubernetes test environment:
* Google Container Engine
* Kubernetes 1.8.1
* Caddy 0.10.10 [FOSS edition](https://hub.docker.com/r/stefanprodan/caddy/)

CLI test environments:
* macOS Siera 10.12.6
* Ubuntu 16.04
* Windows 10 Pro (go 1.9.2 windows/amd64)

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
When we get closer to completing review please can you squash commits?

File: README.md [hunk 1]
```diff
@@ -40,6 +40,8 @@ The main commands supported by the CLI are:
 * `faas-cli deploy` - deploys the functions into a local or remote OpenFaaS gateway
 * `faas-cli remove` - removes the functions from a local or remote OpenFaaS gateway
 * `faas-cli invoke` - invokes the functions and reads from STDIN for the body of the request
+* `faas-cli login` - stores basic auth credentials for OpenFaaS gateways (multi-tenancy support)
```
Comment (alexellis) [inline README.md]: > multi-tenancy support

This isn't the "multi-tenancy support" people on the project are going to expect. Maybe - supports multiple gateways.

File: config/file.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/mitchellh/go-homedir"
+	"gopkg.in/yaml.v2"
+)
+
+var (
```
Comment (alexellis) [inline config/file.go]: Out of curiosity.. is this all original work, or based upon an example?

File: proxy/auth_test.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	"strings"
+
+	"github.com/openfaas/faas-cli/config"
+)
+
+func Test_AuthorizationHeader(t *testing.T) {
+	//setup store
+	config.DefaultFile = "authtest1.yml"
+	gatewayURL := strings.TrimRight("http://openfaas.test/", "/")
+	config.UpdateAuthConfig(gatewayURL, "Aladdin", "open sesame")
```
Comment (alexellis) [inline proxy/auth_test.go]: Have we tested for empty string and edge-cases?

File: vendor.conf [hunk 4]
```diff
@@ -4,4 +4,5 @@ github.com/spf13/cobra 2df9a531813370438a4d79bfc33e21f58063ed87
 github.com/spf13/pflag e57e3eeb33f795204c1ca35f56c44f83227c6e6
 github.com/ryanuber/go-glob 256dc444b735e061061cf46c809487313d5b0065
 github.com/morikuni/aec 39771216ff4c63d11f5e604076f9c45e8be1067b
-github.com/inconshreveable/mousetrap 76626ae9c91c4f2a10f34cad8ce83ea42c93bb75
\ No newline at end of file
+github.com/inconshreveable/mousetrap 76626ae9c91c4f2a10f34cad8ce83ea42c93bb75
+github.com/mitchellh/go-homedir b8bc1bf767474819792c23f32d8286a45736f1c6
```
Comment (alexellis) [inline vendor.conf]: Is this new dep Windows compatible, too?

File: config/file.go (hunk 2 above)
Comment (stefanprodan) [reply config/file.go]: The naming convention follows docker/cli/config and the decode function is using the same escape as in docker registry store, other than that is as "original" as this little pice of code can be :)

File: vendor.conf (hunk 4 above)
Comment (stefanprodan) [reply vendor.conf]: Yes it is see https://github.com/mitchellh/go-homedir/blob/master/homedir.go#L120

File: proxy/auth_test.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) Stefan Prodan 2017. All rights reserved.
```
Comment (alexellis) [inline proxy/auth_test.go]: Is this copied from Lewis PR or visa-versa?

https://github.com/openfaas/faas-cli/pull/180

If so is the Copyright notice appropriate?

File: proxy/auth_test.go (hunk 3 above)
Comment (stefanprodan) [reply proxy/auth_test.go]: The URL empty string validation is done here https://github.com/openfaas/faas-cli/pull/182/files/dc06babac728c33fb128bba45766f229c134a991#diff-4d2d07a462d7a1dda6e42b8c443d6d77R74 and the validity of the URL is tested by calling the gateway `/list` before the URL get's saved in the config.

File: proxy/auth_test.go (hunk 5 above)
Comment (stefanprodan) [reply proxy/auth_test.go]: The hashing comparison it is copied. 

Regarding Copyright wouldn't be better to put "OpenFaaS Team" and not the contributor name? I've seen this approach on other projects. Each commit is signed, so if someone wants to see the actual contributors it can from the commit history.

File: proxy/auth_test.go (hunk 3 above)
Comment (alexellis) [reply proxy/auth_test.go]: My question was - have we got a unit test?

File: proxy/auth_test.go (hunk 5 above)
Comment (alexellis) [reply proxy/auth_test.go]: If you like you can put OpenFaaS authors on your new commits in this PR. Unlikely to retrofit across all repos at this point.

File: proxy/auth_test.go (hunk 3 above)
Comment (Lewiscowles1986) [reply proxy/auth_test.go]: I'm also unsure we should be using username password. My original PR only did it because it incorrectly used cli-arguments. We are storing base64, why not add the requisite header?

Comment (stefanprodan) [reply proxy/auth_test.go]: testes added for empty string and invalid URLs in https://github.com/openfaas/faas-cli/pull/182/commits/35df33c431d82e5137dfd0928c48940d185d2170

**Comment by stefanprodan** [comment]
@alexellis you can squash all commits at merge, see https://github.com/blog/2141-squash-your-commits

File: vendor/github.com/mitchellh/go-homedir/LICENSE [hunk 6]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
Comment (nenadilic84) [inline vendor/github.com/mitchellh/go-homedir/LICENSE]: @alexellis, @stefanprodan, I don't like the fact that we checkin the code in vendor. I think I mentioned this before, but it should be that only use the vendor.conf to have the right versions, and then let the tool checkout everything before building. I would just put .gitignore in this folder.

Comment (alexellis) [reply vendor/github.com/mitchellh/go-homedir/LICENSE]: This is how dependencies are managed on Golang projects. It may seem odd, but you can find more information by looking at Docker or Google's Go projects.

**Comment by alexellis** [comment]
@stefanprodan I won't be altering the commits at merge time because it'll lose the sign off. If you could do that when the work is ready that would work better.

**Comment by stefanprodan** [comment]
Squash done and ready for merge :)

**Comment by alexellis** [comment]
Thanks for squashing. I'll try to do a detailed code-review over the weekend. This is great timing.

I've done some testing on my local machine with basic auth on nginx and so far looks very good :+1:

File: config/file.go (hunk 2 above)
Comment (alexellis) [reply config/file.go]: Can you link me to that code please? Thank you

**Comment by alexellis** [comment]
@johnmccabe when you're up to it, please can you review the config changes?

File: proxy/invoke.go [hunk 7]
```diff
@@ -26,13 +27,13 @@ func InvokeFunction(gateway string, name string, bytesIn *[]byte, contentType st
...
 	if qsErr != nil {
 		return nil, qsErr
-
 	}
 
 	gatewayURL := gateway + "/function/" + name + qs
 	// fmt.Println(gatewayURL)
 	req, _ := http.NewRequest(http.MethodPost, gatewayURL, reader)
 	req.Header.Add("Content-Type", contentType)
+	SetAuth(req, gateway)
```
Comment (johnmccabe) [inline proxy/invoke.go]: This is good for now.

But I think we will want to be more nuanced with locking down endpoints, with functions and system being secured using different credentials (and most likely with individual functions with their own auth (or none)). (^ @alexellis)

File: proxy/auth_test.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) OpenFaaS authors 2017. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package proxy
+
+import (
+	"net/http"
+	"testing"
+
```
Comment (johnmccabe) [inline proxy/auth_test.go]: Delete this line

File: config/file.go [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+	conf := &ConfigFile{
+		AuthConfigs: make(map[string]AuthConfig),
+		FilePath:    filePath,
+	}
+
+	return conf, nil
+}
+
+// EnsureFile creates the root dir and config file
+func EnsureFile() (string, error) {
```
Comment (johnmccabe) [inline config/file.go]: Would be cleaner when testing to allow the homedir to be overridden, currently you get your local `~/.openfaas/` dir polluted with test YAML.

File: config/file_test.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+	"os"
+	"regexp"
+	"strings"
+	"testing"
+)
+
+func Test_LookupAuthConfig_WithNoConfigFile(t *testing.T) {
+	DefaultFile = "test1.yml"
+	_, _, err := LookupAuthConfig("http://openfaas.test1")
+
```
Comment (johnmccabe) [inline config/file_test.go]: Can you get rid of the extra line between the function calls and error handling, same for the rest of this file.

File: proxy/invoke.go (hunk 7 above)
Comment (alexellis) [reply proxy/invoke.go]: Sending basic authentication when the function is open will result in the header being ignored, so we should be ok.

File: commands/login.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+func validateLogin(url string, user string, pass string) error {
+	// TODO: provide --insecure flag for this
+	tr := &http.Transport{
+		DisableKeepAlives: true,
+		TLSClientConfig:   &tls.Config{InsecureSkipVerify: true},
+	}
+	client := &http.Client{
+		Transport: tr,
+		Timeout:   time.Duration(5 * time.Second),
```
[the remark this answers is not in the document]
Comment (alexellis) [reply commands/login.go]: Agreed

**Comment by stefanprodan** [comment]
@johnmccabe @alexellis I've addressed all your change requests. After this gets merged I will do the `logout` command.

File: README.md [hunk 12]
```diff
@@ -49,9 +52,9 @@ You can chose between using a [programming language template](https://github.com
 
 **Templates**
 
-Command: `faas-cli new FUNCTION_NAME --lang python/node/ruby/Dockerfile/etc`
+Command: `faas-cli new FUNCTION_NAME --lang python/node/go//ruby/Dockerfile/etc`
```
Comment (ericstoekl) [inline README.md]: Are there supposed to be two `/` characters after `go`?

Comment (johnmccabe) [reply README.md]: I think you're looking at an old revision there @ericstoekl ?

**Comment by johnmccabe** [comment]
Just for future reference, it would be much easier to not squash after committing review updates as I have to review the entire PR again now :(

File: commands/login.go [hunk 13]
```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Stefan Prodan 2017. All rights reserved.
```
[the remark this answers is not in the document]
Comment (alexellis) [reply commands/login.go]: Yes needs to be assigned to the project for new files - OpenFaaS authors.

[the remark this answers is not in the document]
Comment (johnmccabe) [reply commands/login.go]: @stefanprodan can you update this per - https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md#copyright-notice

File: config/file.go [hunk 14]
```diff
@@ -0,0 +1,222 @@
+// Copyright (c) Stefan Prodan 2017. All rights reserved.
```
Comment (johnmccabe) [inline config/file.go]: Can you update this per - https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md#copyright-notice

File: config/file_test.go [hunk 15]
```diff
@@ -0,0 +1,157 @@
+// Copyright (c) Stefan Prodan 2017. All rights reserved.
```
Comment (johnmccabe) [inline config/file_test.go]: Can you update this per - https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md#copyright-notice

File: proxy/auth.go [hunk 16]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) OpenFaaS authors 2017. All rights reserved.
```
Comment (johnmccabe) [inline proxy/auth.go]: Can you update this per - https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md#copyright-notice

File: proxy/auth_test.go [hunk 17]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) OpenFaaS authors 2017. All rights reserved.
```
Comment (johnmccabe) [inline proxy/auth_test.go]: Can you update this per - https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md#copyright-notice

File: config/file_test.go [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+
+func Test_New_NoFile(t *testing.T) {
+	_, err := New("")
+	if err == nil {
+		t.Error("expected to fail on empty file path")
+	}
+}
+
+func Test_EnsureFile(t *testing.T) {
+	DefaultDir = "~/.openfaas_test"
```
Comment (johnmccabe) [inline config/file_test.go]: Can you use `ioutil.TempDir("", "faas-cli-file-test")` here so we don't pollute the users directory.

File: config/file.go (hunk 2 above)
Comment (johnmccabe) [reply config/file.go]: @alexellis do you still need this?

**Comment by johnmccabe** [comment]
Its a 👍 from me, just need to get these two bits sorted and I think its it looks good to merge:
- typo - https://github.com/openfaas/faas-cli/pull/182#discussion_r147745082
- question from @alexellis - https://github.com/openfaas/faas-cli/pull/182#discussion_r147549275

File: config/file.go (hunk 2 above)
Comment (stefanprodan) [reply config/file.go]: Sorry I've missed this question. Here it is https://github.com/docker/cli/blob/master/cli/config/configfile/file.go

**Comment by stefanprodan** [comment]
@johnmccabe that typo got into my PR from a master merge. I will fix it.

File: commands/login.go [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+	username      string
+	password      string
+	passwordStdin bool
+)
+
+func init() {
+	loginCmd.Flags().StringVar(&gateway, "gateway", defaultGateway, "Gateway URI")
+	loginCmd.Flags().StringVarP(&username, "username", "u", "", "Gateway username")
+	loginCmd.Flags().StringVarP(&password, "password", "p", "", "Gateway password")
+	loginCmd.Flags().BoolVar(&passwordStdin, "password-stdin", false, "Reads the gateway password from stdin")
```
Comment (alexellis) [inline commands/login.go]: We can tweak the UX after merging but password-stdin should be as easy to use as -p so that people don't just default to having their password in the bash history.

File: commands/login.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (alexellis) [reply commands/login.go]: @stefanprodan ..  something like?

```
Validating login... OK.
Validating login... Failed.
```

File: config/file.go (hunk 9 above)
Comment (alexellis) [reply config/file.go]: Yes - talked about this with @johnmccabe on Sunday. We should be using os.Tempdir or similar for integration tests.

As a side note about the cli tests, integration tests should perhaps be in their own area so we can run them independently.

Comment (stefanprodan) [reply config/file.go]: I've used `os.Tempdir` take a look at the latest commit.

File: config/file_test.go [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+	DefaultDir, _ = ioutil.TempDir("", "faas-cli-file-test")
+	DefaultFile = "test1.yml"
+	_, _, err := LookupAuthConfig("http://openfaas.test1")
+	if err == nil {
+		t.Errorf("Error was not returned")
+	}
+
+	r := regexp.MustCompile(`(?m:config file not found)`)
+	if !r.MatchString(err.Error()) {
+		t.Errorf("Error not matched: %s", err.Error())
```
Comment (alexellis) [inline config/file_test.go]: I don't like the pattern we have in these tests (it's not from Stefan, but from some historic tests that got copied) `Error not matched` is super vague.. can we do better? What would you suggest guys? @stefanprodan @johnmccabe 

Imagine reading this deep in an CI log.

File: config/file_test.go [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+	if err != nil {
+		t.Errorf("got error %s", err.Error())
+	}
+
+	if user != u || pass != p {
+		t.Errorf("got user %s and pass %s, expected %s %s", user, pass, u, p)
+	}
+}
+
+func Test_UpdateAuthConfig_Update(t *testing.T) {
```
Comment (alexellis) [inline config/file_test.go]: Thank you for taking the time to add the test coverage. This means a lot @stefanprodan

File: proxy/auth.go [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+package proxy
+
+import (
+	"net/http"
+
+	"github.com/openfaas/faas-cli/config"
+)
+
+//SetAuth sets basic auth for the given gateway
+func SetAuth(req *http.Request, gateway string) {
```
Comment (alexellis) [inline proxy/auth.go]: This function feels orphaned to me. I know why it's here because it gives a sense of removing duplication, but I think we could do this in a cleaner way. I saw a similar pattern for user-agent and didn't like it either.

I'm thinking of our own request factory which adds the things we know we need.

Doesn't have to be part of this PR. @stefanprodan  @rgee0 - thoughts?

File: commands/login.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (stefanprodan) [reply commands/login.go]: I've done this, see line 78

File: proxy/auth_test.go [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+		t.Errorf("got header %q, want %q", header, expected)
+	}
+}
+
+func Test_SetAuth_SkipAuthorization(t *testing.T) {
+	//setup store
+	config.DefaultDir, _ = ioutil.TempDir("", "faas-cli-auth-test")
+	config.DefaultFile = "authtest2.yml"
+	gatewayURL := strings.TrimRight("http://openfaas.test/", "/")
+	config.UpdateAuthConfig(gatewayURL, "Aladdin", "open sesame")
```
Comment (alexellis) [inline proxy/auth_test.go]: I like this test a lot but I had to read it 3-4 times. I think it could be easier to read if we broke out the URLs into variables so it's easier to read and understand at a glance.

```
basicAuthURL
openURL
```

File: proxy/deploy.go [hunk 24]
```diff
@@ -89,6 +90,8 @@ func DeployFunction(fprocess string, gateway string, functionName string, image
 
 		deployedURL := fmt.Sprintf("URL: %s/function/%s\n", gateway, functionName)
 		fmt.Println(deployedURL)
+	case 401:
```
Comment (alexellis) [inline proxy/deploy.go]: Please can you update to http.StatusUnauthorized etc?

File: proxy/delete.go [hunk 25]
```diff
@@ -39,6 +40,8 @@ func DeleteFunction(gateway string, functionName string) {
 		fmt.Println("Removing old service.")
 	case 404:
 		fmt.Println("No existing service to remove")
+	case 401:
```
Comment (alexellis) [inline proxy/delete.go]: Please can you update to http.StatusUnauthorized etc? I know these have the numbers in (that's on me..) but let's fix while we're in there.

File: proxy/list.go [hunk 26]
```diff
@@ -46,7 +49,8 @@ func ListFunctions(gateway string) ([]requests.Function, error) {
 		if jsonErr != nil {
 			return nil, fmt.Errorf("cannot parse result from OpenFaaS on URL: %s\n%s", gateway, jsonErr.Error())
 		}
-
+	case 401:
```
Comment (alexellis) [inline proxy/list.go]: Please go over to http.StatusEtc...

File: README.md [hunk 27]
```diff
@@ -40,6 +40,8 @@ The main commands supported by the CLI are:
 * `faas-cli deploy` - deploys the functions into a local or remote OpenFaaS gateway
 * `faas-cli remove` - removes the functions from a local or remote OpenFaaS gateway
 * `faas-cli invoke` - invokes the functions and reads from STDIN for the body of the request
+* `faas-cli login` - stores basic auth credentials for OpenFaaS gateway (supports multiple gateways)
```
[the remark this answers is not in the document]
Comment (stefanprodan) [reply README.md]: I've implemented `logout` and added tests for store remove func

**Comment by alexellis** [comment]
When we spoke on Slack we settled on `logout` being out of scope for this PR. I'll have to set some more time aside to review.

**Comment by stefanprodan** [comment]
The `logout` code is straight forward, 20 lines of code with 100% test coverage :)

File: commands/login.go [hunk 28]
```diff
@@ -0,0 +1,140 @@
...
+		if len(username) == 0 {
+			return errors.New("must provide --username with --password-stdin")
+		}
+
+		passwordStdin, err := ioutil.ReadAll(os.Stdin)
+		if err != nil {
+			return err
+		}
+
+		password = strings.TrimSpace(string(passwordStdin))
```
Comment (alexellis) [inline commands/login.go]: Do we also need to trim the --password flag data?

**Comment by alexellis** [comment]
Started to do some testing on this. I found the output format of ~/.openfaas/config.yml looked a little odd:

```
auths:
  http://127.0.0.1:8081:
    auth: basic
    token: value
  http://other-site:8080:
    auth: basic
    token: value
  http://other-gateway:8081:
    auth: basic
    token: value
```

There was a comment Joe Beda made last Friday about using values for keys which seemed to make sense at the time. We have broken the rule for function names in the OpenFaaS stack YAML but it looks worse here with the URL and then the final colon on the line.

Do you have any other suggestions on improving the key for the URL?

**Comment by stefanprodan** [comment]
@alexellis Regarding the config format, I am using the same format as Docker CLI uses for storing the registries auth. I don't see a problem with it, this is how a golang map is getting serialized in yaml.

File: proxy/auth.go (hunk 22 above)
Comment (alexellis) [reply proxy/auth.go]: Would like your thoughts.

File: proxy/delete.go [hunk 29]
```diff
@@ -40,10 +41,12 @@ func DeleteFunction(gateway string, functionName string) error {
 	}
 
 	switch delRes.StatusCode {
-	case 200, 201, 202:
+	case http.StatusOK, http.StatusCreated, http.StatusAccepted:
```
Comment (alexellis) [inline proxy/delete.go]: Thanks for making this better.

**Comment by alexellis** [comment]
Have you tested thoroughly on Windows?

**Comment by alexellis** [comment]
I think this format is closer to the "Kubernetes way of YAML" - https://github.com/estesp/bucketbench

```
name: Basic
image: alpine:latest
command: date
rootfs: /home/estesp/containers/alpine
detached: true
drivers:
  - 
   type: Docker
   threads: 5
   iterations: 15
  - 
   type: Runc
   threads: 5
   iterations: 50
```

I don't know how easy it is to serialize that way.

Another example (looking at the "containers")

```
apiVersion: apps/v1beta2 # for versions before 1.7.0 use apps/v1beta1
kind: Deployment
metadata:
  name: nginx-deployment
spec:
  replicas: 3
  template:
    metadata:
      labels:
        app: nginx
    spec:
  - name: nginx-container
    image: nginx

  - name: debian-container
    image: debian
    command: ["/bin/sh"]
```

Both appear to be applying an array of "map/object" in order to force this syntax.. I mean you can checkout the video from Joe. I think we should at least evaluate his advice. Around here - https://www.youtube.com/watch?v=iMzf_oWsRi0

File: commands/login.go [hunk 30]
```diff
@@ -0,0 +1,140 @@
...
+	if res.Body != nil {
+		defer res.Body.Close()
+	}
+
+	if res.TLS == nil {
+		fmt.Println("WARNING! Communication is not secure, please consider using HTTPS. Letsencrypt.org offers free SSL/TLS certificates.")
+	}
+
+	switch res.StatusCode {
+	case 200:
```
Comment (alexellis) [inline commands/login.go]: More numbers, but we can sort these later

**Comment by alexellis** [comment]
Minor nit between `docker login` and `faas-cli login`:

```
docker login -u alex
(starts reading password from stdin)
```

```
faas-cli login -u alex
Error: password cannot be an empty string
```

Might be worth looking at how we do the `invoke` command and doing the same. Could be a follow-up PR / issue.

**Comment by stefanprodan** [comment]
I've updated the description, the CLI unit tests and login/logout manual testing were performed on:

* macOS Siera 10.12.6
* Ubuntu 16.04
* Windows 10 Pro (go 1.9.2 windows/amd64)

**Comment by stefanprodan** [comment]
@alexellis regarding the config, if you look at bucketbench you'll see that the drivers are an array not a map. You can't force a map to be serialized as an array, so you'll not get the `-`.

**Comment by rgee0** [comment]
My sense is that the yaml format should be formed of nested maps so that each value has a name. This enhances the readability and use ability for the user.

It’s actually not too far off, probably just needs the gateway uri to identify itself as a gateway value.

**Comment by stefanprodan** [comment]
A nested map is not an array, if @alexellis wants the config to look like the k8s one I need to change it to an array.

```yml
auths:
 - gateway: http://127.0.0.1:8081
    auth: basic
    token: value
 - gateway: http://other-site:8080
    auth: basic
    token: value
```

@alexellis should I rewrite the config package to have the above format?

File: config/file_test.go (hunk 20 above)
Comment (johnmccabe) [reply config/file_test.go]: I tend to use table drive as much as possible with the name of the testing being printed and for assertion failures printing the both the expected and actual values, kinda like:

    t.Errorf("Returned error [%v] does not match expected value [%v]", err, expectedErr)

@stefanprodan as an aside, you should be using `%v` with the raw error rather than `%s` and calling the `.String()` function.

**Comment by johnmccabe** [comment]
Regarding the file structure:
- Is the config file intended to be human readable/editable?
  - if it is, then yes to named keys
- if not then using a map is easier as you don’t need to iterate through the slice to find a particular gateway

**Comment by stefanprodan** [comment]
I don't see users editing the auth data in the config file, but if the config will grow in time I think we should make it as readable as possible. Changing the code from map to array is an easy change with little overhead. @alexellis what's your take on this?

File: commands/login.go (hunk 28 above)
Comment (stefanprodan) [reply commands/login.go]: done in the latest commit

**Comment by stefanprodan** [comment]
I've switched from map to array, this how it looks:

```yml
auths:
- gateway: http://35.198.149.1
  auth: basic
  token: value
- gateway: http://35.198.149.2
  auth: basic
  token: value
```

**Comment by developius** [comment]
This is really cool. Clearly this PR has been merged but I had a comment about using `--password-stdin` though - all the CLIs I've ever used use the format `-p -` (or `--password -`) to mean "grab this from stdin". 
Example:

```
$ cat ~/faas_pass.txt | faas-cli login -u user -p - --gateway http://localhost:8080
```
Thoughts?

**Comment by alexellis** [comment]
What does the Docker CLI do?

**Comment by developius** [comment]
It uses `--password-stdin`, but that's the only occurrence I've ever seen of not using `-p -`

**Comment by stefanprodan** [comment]
@developius how will you do that with cobra?

**Comment by developius** [comment]
@stefanprodan I think you could just check to see if the password flag was set to `-`, and if it was do the same logic as if `passwordStdin` was `true` [login.go#L61-L72](https://github.com/stefanprodan/faas-cli/blob/ee859a3e699160a17d23ac05f33adcb41e1e8df9/commands/login.go#L61-L72)

**Comment by stealthybox** [comment]
`-` is usually a program specific alias for `/dev/stdin`.
( I actually quite dislike this because it overrides other shell syntax, but it's fairly standard. )

For `-p` the param isn't a file name.
I believe the proper flag for a `-` alias would be something like `--password-file`.

This would be useful for other things like secrets in a tmpfs.
